File: problems/unicycle_traj_tracking/data/physical_experiment/draw.py
import numpy as np
import matplotlib.pyplot as plt
import scipy.io as sio
import os
from problems.unicycle_traj_tracking.data.butterfly_tracking.draw import load_data_and_save_figure

def main():
    gmpc_solver_time = np.load('Geomtric_model_predictive_control_store_solve_time.npy')
    nmpc_solver_time = np.load('Nonlinear_model_predictive_control_store_solve_time.npy')
    # save as mat file
    sio.savemat('solve_time.mat', {'gmpc_solver_time': gmpc_solver_time, 'nmpc_solver_time': nmpc_solver_time})
    # calculate mean and std
    gmpc_solver_time_mean = np.mean(gmpc_solver_time)
    gmpc_solver_time_std = np.std(gmpc_solver_time-gmpc_solver_time_mean)
    nmpc_solver_time_mean = np.mean(nmpc_solver_time)
    nmpc_solver_time_std = np.std(nmpc_solver_time-nmpc_solver_time_mean)
    # max and min
    gmpc_solver_time_max = np.max(gmpc_solver_time)
    gmpc_solver_time_min = np.min(gmpc_solver_time)
    nmpc_solver_time_max = np.max(nmpc_solver_time)
    nmpc_solver_time_min = np.min(nmpc_solver_time)
    print('gmpc_solver_time_mean: ', gmpc_solver_time_mean)
    print('gmpc_solver_time_std: ', gmpc_solver_time_std)
    print('nmpc_solver_time_mean: ', nmpc_solver_time_mean)
    print('nmpc_solver_time_std: ', nmpc_solver_time_std)
    print('gmpc_solver_time_max: ', gmpc_solver_time_max)
    print('gmpc_solver_time_min: ', gmpc_solver_time_min)
    print('nmpc_solver_time_max: ', nmpc_solver_time_max)
    print('nmpc_solver_time_min: ', nmpc_solver_time_min)

    plt.figure()
    plt.boxplot([gmpc_solver_time, nmpc_solver_time], labels=['GMPC', 'NMPC'],showfliers=False, patch_artist=True )
    plt.ylabel('Solve time (s)')
    plt.tight_layout()
    plt.savefig('solve_time.jpg')
    plt.show()

if __name__ == '__main__':
    main()

File: problems/unicycle_traj_tracking/data/physical_experiment/bag/solvetime/draw.m
solve_time = load('solve_time.mat');
gmpc_solver_time = solve_time.gmpc_solver_time;
nmpc_solver_time = solve_time.nmpc_solver_time;

% calculate mean and std
gmpc_solver_time_mean = mean(gmpc_solver_time);
gmpc_solver_time_std = std(gmpc_solver_time - gmpc_solver_time_mean);
nmpc_solver_time_mean = mean(nmpc_solver_time);
nmpc_solver_time_std = std(nmpc_solver_time - nmpc_solver_time_mean);

% max and min
gmpc_solver_time_max = max(gmpc_solver_time);
gmpc_solver_time_min = min(gmpc_solver_time);
nmpc_solver_time_max = max(nmpc_solver_time);
nmpc_solver_time_min = min(nmpc_solver_time);

fprintf('gmpc_solver_time_mean: %f\n', gmpc_solver_time_mean);
fprintf('gmpc_solver_time_std: %f\n', gmpc_solver_time_std);
fprintf('nmpc_solver_time_mean: %f\n', nmpc_solver_time_mean);
fprintf('nmpc_solver_time_std: %f\n', nmpc_solver_time_std);
fprintf('gmpc_solver_time_max: %f\n', gmpc_solver_time_max);
fprintf('gmpc_solver_time_min: %f\n', gmpc_solver_time_min);
fprintf('nmpc_solver_time_max: %f\n', nmpc_solver_time_max);
fprintf('nmpc_solver_time_min: %f\n', nmpc_solver_time_min);

h= figure;

labels = {'GMPC', 'NMPC'};

font_size = 18;
line_width = 2;
set(h, 'DefaultTextFontSize', font_size);
boxplot([gmpc_solver_time', nmpc_solver_time'], 'Labels', {'GMPC', 'NMPC'}, 'Whisker', 1, 'Colors', 'b', 'symbol','');
% Set the fill color for each box
h_box=findobj(gca,'Tag','Box');
h_M = findobj(gca,'Tag','Median');

color(1,:) = [255 5 5]/255;
color(2,:) = [255 100 5]/255;
color(3,:) = [255 255 5]/255;
color(4,:) = [200 255 5]/255;
color(5,:) = [5 255 5]/255;
color(6,:) = [5 255 100]/255;
color(7,:) = [5 255 255]/255;
color(8,:) = [5 200 255]/255;
color(9,:) = [5 5 255]/255;
color(10,:) = [255 5 255]/255;
for j=1:length(h_box)
    h_M(j).Color='k';
    uistack(patch(get(h_box(j),'XData'),get(h_box(j),'YData'),color(6),'FaceAlpha',1,'linewidth',1.5),'bottom');
end
ylim([0.012  0.035])
ylabel('Solve time (s)');
set(gca,'XTick',1:2)%2 because only exists 2 boxplot
set(gca,'XTickLabel',{'GMPC', 'NMPC'},'FontSize',14)
% tightfig;
% saveas(gcf, 'solve_time.jpg');
% close(gcf);